### DISCUSSION TITLE
Packaging for archlinux (aur)

### DISCUSSION THREAD

**Discussion by mmdbalkhi** [spenserblack/repofetch#228] [Development]
Hello! I hope you are well. 

I liked this project and wanted to package it for [aur](https://aur.archlinux.org/). can I do this?

**Comment by spenserblack** [comment]
Yes, I'm happy you're taking an interest in this project!

Just to confirm: would you be packaging the new version written in Ruby, or the old version (`< 0.4.0`) written in Rust?

Also, just an FYI that I'm going to release `0.4.1` soon to fix [an embarrassing bug](https://github.com/spenserblack/repofetch/issues/225).

Let me know when you're done, and I'll add you to the All Contributors file :smiley: Also, I'd appreciate a PR with those installation instructions once it's available :slightly_smiling_face:

**Reply by mmdbalkhi** [reply]
Thank you!

> Just to confirm: would you be packaging the new version written in Ruby, or the old version (< 0.4.0) written in Rust?

I want to package the code written in Ruby, but I can also package the Rust version if you want.

**Reply by mmdbalkhi** [reply]
ruby-repofetch: https://aur.archlinux.org/packages/ruby-repofetch

**Reply by spenserblack** [reply]
(let's see if this works in discussions)

@allcontributors please add @mmdbalkhi for platform

**Reply by spenserblack** [reply]
>I want to package the code written in Ruby, but I can also package the Rust version if you want.

I'm OK with only the Ruby version being packaged :+1:. I'd like this to be the version that everyone uses, and I'll prioritize work for the Ruby version over the Rust version.

**Reply by spenserblack** [reply]
Still getting used to allcontributors, maybe [the hyphen works :sweat_smile:](https://allcontributors.org/docs/en/bot/usage). If it doesn't work at all in discussions, I'll wait for your PR :slightly_smiling_face:

@all-contributors please add @mmdbalkhi for platform

**Reply by mmdbalkhi** [reply]
here you are: 
- #230

**Reply by mmdbalkhi** [reply]
Is there a binary version of the project? The current package is built from source and I think it is better to have a stable package based on binary.

**Reply by mmdbalkhi** [reply]
@spenserblack  github-cli is optional?

**Reply by spenserblack** [reply]
>github-cli is optional?

Yes, this app doesn't require `gh`. `gh auth token` is only used in the dotenv example to illustrate that [dotenv supports bash substitutions](https://github.com/bkeepers/dotenv). One could also simply create a token from the web and copy and paste it to `repofetch.env`. Perhaps that example is confusing if it makes it look like `gh` is required :thinking:

>Is there a binary version of the project?

I know nothing about packaging for AUR, and this is my first real Ruby project, so I still have a lot to learn, and I probably can't answer this question adequately right now :laughing: 
I would have to research some options, but right now I would say no. The closest I could get to a binary would probably be to build the `.gem` file and include it as a release asset. Would a `repofetch.gem` asset that can be downloaded from github releases work? You would still need to `gem install` the file.

**Reply by spenserblack** [reply]
[`repofetch.gem` should now get published with each release](https://github.com/spenserblack/repofetch/releases/tag/v0.4.2-rc.1). You should be able to download it from https://github.com/spenserblack/repofetch/releases/latest/download/repofetch.gem the next time I make a release that isn't a prerelease.

**Reply by mmdbalkhi** [reply]
It looks useful. I have a few questions:

- Does this `repofetch.gem` file work for any kind of system architecture?
- Can you `tar gz` the MD and license files along with the `repofetch.gem` file?

Edit:
If you agree, I will submit a PR soon about the second one.

**Reply by spenserblack** [reply]
It's interpreted code, so it will work anywhere that `ruby` would work (Ruby 2.7, 3.0, and 3.1 are supported).

We can certainly `tar gz` those files, but I think I should take a look at some popular Ruby projects that are packaged for archlinux, to see what others might do.

Note that the `.gem` file actually is an archive: you can run `tar xf repofetch.gem` to get the `data.tar.gz` file, then extract that `data` file, and `LICENSE` and `*.md` will be there, because any uppercase file is included here: https://github.com/spenserblack/repofetch/blob/b01827b095dc8d9fd09371302175d31b87ea3f81/repofetch.gemspec#L18

The files will also be where the gem is installed, `<install-path>/gems/repofetch-<version>/LICENSE`.

**Reply by mmdbalkhi** [reply]
Thanks for the helpful information

> We can certainly tar gz those files, but I think I should take a look at some popular Ruby projects that are packaged for archlinux, to see what others might do.

I agree with your opinion. The current package is based on the ruby-matrix package.

I checked now, it seems that the version should not **contain colons**, **forward slashes**, **hyphens** or **whitespace**. I pushed the source code of package [here](https://github.com/mmdbalkhi/pkgbuilds/blob/main/ruby-repofetch-bin/PKGBUILD), ~~but I will wait to push into aur until the next update version name is fixed.~~

**Reply by spenserblack** [reply]
So this is a library, not an executable (it's actually a dependency of repofetch :laughing:), but perhaps [`octokit.rb`](https://archlinux.org/packages/community/any/ruby-octokit/) could be an example. It looks like [this is the `PKGBUILD`](https://github.com/archlinux/svntogit-community/blob/2ff0ab1c7aacdd828ecf4eb5a2703f64e9ae0364/ruby-octokit/repos/community-any/PKGBUILD).

**Reply by mmdbalkhi** [reply]
haha :joy: but there is no difference in their installation. Anyway. can we download `repofetch.gem` from [rubygems.org](https://rubygems.org/gems/repofetch)? I put this in the package, which means that there is no need to upload the `repofetch.gem` file to the releases.

**Reply by spenserblack** [reply]
I suppose you could download from rubygems.org, if that's acceptable for aur. Otherwise, I think you can make the `source` `https://github.com/spenserblack/repofetch/archive/v$pkgver/$pkgname-$pkgver.tar.gz` (I'm actually surprised to find out that https://github.com/spenserblack/repofetch/archive/v0.4.1/ruby-repofetch-0.4.1.tar.gz works! Even https://github.com/spenserblack/repofetch/archive/v0.4.1/foo.tar.gz will download `repofetch-0.4.1.tar.gz`)

**Reply by mmdbalkhi** [reply]
ruby-repofetch-bin:
https://aur.archlinux.org/packages/ruby-repofetch-bin

**Reply by spenserblack** [reply]
> So this is a library, not an executable (it's actually a dependency of repofetch )

Oh, when I said "this" and "it's" I was referring to `octokit.rb`, not `repofetch`. Sorry if I caused any confusion by my wording.

**Reply by mmdbalkhi** [reply]
> > So this is a library, not an executable (it's actually a dependency of repofetch )
> 
> Oh, when I said "this" and "it's" I was referring to `octokit.rb`, not `repofetch`. Sorry if I caused any confusion by my wording.

No, I understand what you mean. That's why I said "there is no difference in their installation". I think Ruby moves the library and the executable file to their correct place. And fortunately, the [binary package](https://aur.archlinux.org/packages/ruby-repofetch-bin) worked successfully.

**Reply by spenserblack** [reply]
Great! Thanks for all your hard work! So I guess the installation instruction needs a slight update, right?

**Reply by mmdbalkhi** [reply]
Yes, I'll submit a PR as soon.

**Reply by mmdbalkhi** [reply]
@spenserblack Sorry for being late.

- #236 

If there is any question or problem, please message me in this discussion or on Discord (`mmdbalkhi#1053`) if you have it.
